mypy/checkexpr.py

```diff
@@ -3505,6 +3508,63 @@ def visit_op_expr(self, e: OpExpr) -> Type:
         else:
             raise RuntimeError(f"Unknown operator {e.op}")
 
+    def literal_value_from_expr(
+        self, expr: Expression, typ: Type | None = None
+    ) -> tuple[list[str | int], str, bool] | None:
+        if isinstance(expr, StrExpr):
+            return [expr.value], "builtins.str", False
+        if isinstance(expr, IntExpr):
+            return [expr.value], "builtins.int", False
+        if isinstance(expr, BytesExpr):
+            return [expr.value], "builtins.bytes", False
+
+        typ = typ or self.accept(expr)
```

Comment:
Calling `accept` here can be inefficient, since the caller may also call `accept` on the expression at some point, causing duplicate work. This could cause exponential time complexity in nested expression trees (we already have examples of this and it can cause extremely slow type checking). Maybe the caller can pass the expression type, so that we only call `accept` once per expression? Maybe the new logic should be called from `check_op`, when we have the right operand type?